src/main/drivers/rx/rx_sx127x.c

```diff
@@ -0,0 +1,471 @@
+        return 62.5E3;
+    case SX127x_BW_125_00_KHZ:
+        return 125E3;
+    case SX127x_BW_250_00_KHZ:
+        return 250E3;
+    case SX127x_BW_500_00_KHZ:
+        return 500E3;
+    }
+    return -1;
+}
+
+uint32_t sx127xGetCurrBandwidthNormalisedShifted(const sx127x_bandwidth_e bw) // this is basically just used for speedier calc of the freq offset, pre compiled for 32mhz xtal
+{
+    switch (bw) {
+    case SX127x_BW_7_80_KHZ:
+        return 1026;
+    case SX127x_BW_10_40_KHZ:
+        return 769;
+    case SX127x_BW_15_60_KHZ:
+        return 513;
+    case SX127x_BW_20_80_KHZ:
+        return 385;
+    case SX127x_BW_31_25_KHZ:
+        return 256;
+    case SX127x_BW_41_70_KHZ:
+        return 192;
+    case SX127x_BW_62_50_KHZ:
+        return 128;
+    case SX127x_BW_125_00_KHZ:
+        return 64;
+    case SX127x_BW_250_00_KHZ:
+        return 32;
+    case SX127x_BW_500_00_KHZ:
+        return 16;
+    }
+    return -1;
+}
+
+void sx127xSetPPMoffsetReg(const int32_t offset, const uint32_t freq)
+{
+    int8_t offsetPPM = (offset * 1e6 / freq) * 95 / 100;
+    sx127xWriteRegister(SX127x_PPMOFFSET, (uint8_t) offsetPPM);
+}
+
+static bool sx127xGetFrequencyErrorbool(void)
+{
+    return (sx127xReadRegister(SX127X_REG_FEI_MSB) & 0x08) >> 3; // returns true if pos freq error, neg if false
+}
+
+int32_t sx127xGetFrequencyError(const sx127x_bandwidth_e bw)
+{
+    uint8_t reg[3] = {0x0, 0x0, 0x0};
+    sx127xReadRegisterBurst(SX127X_REG_FEI_MSB, reg, sizeof(reg));
+
+    uint32_t regFei = ((reg[0] & 0x07) << 16) + (reg[1] << 8) + reg[2];
+
+    int32_t intFreqError = regFei;
+
+    if ((reg[0] & 0x08) >> 3) {
+        intFreqError -= 524288; // Sign bit is on
+    }
+
+    int32_t fErrorHZ = (intFreqError >> 3) * (sx127xGetCurrBandwidthNormalisedShifted(bw)); // bit shift hackery so we don't have to use floaty bois; the >> 3 is intentional and is a simplification of the formula on page 114 of sx1276 datasheet
+    fErrorHZ >>= 4;
+
+    return fErrorHZ;
+}
+
+void sx127xAdjustFrequency(int32_t offset, const uint32_t freq)
+{
+    if (sx127xGetFrequencyErrorbool()) { //logic is flipped compared to original code
+        if (offset > SX127x_FREQ_CORRECTION_MIN) {
+            offset -= 1;
+        } else {
+            offset = 0; //reset because something went wrong
+        }
+    } else {
+        if (offset < SX127x_FREQ_CORRECTION_MAX) {
+            offset += 1;
+        } else {
+            offset = 0; //reset because something went wrong
+        }
+    }
+    sx127xSetPPMoffsetReg(offset, freq); //as above but corrects a different PPM offset based on freq error
+}
+
+uint8_t sx127xUnsignedGetLastPacketRSSI(void)
+{
+    return sx127xGetRegisterValue(SX127X_REG_PKT_RSSI_VALUE, 7, 0);
+}
+
+int8_t sx127xGetLastPacketRSSI(void)
+{
+    return (-157 + sx127xGetRegisterValue(SX127X_REG_PKT_RSSI_VALUE, 7, 0));
+}
+
+int8_t sx127xGetCurrRSSI(void)
+{
+    return (-157 + sx127xGetRegisterValue(SX127X_REG_RSSI_VALUE, 7, 0));
+}
```

Comment:
I'm not familiar with this chip, but is int8_t adequate here?

Reply:
yes, this is in dBm which can be negative. Actual register value is positive, but it is offset by -157 in line 414.